PKGBUILD for snpeff ver 5.1

kbipinkumar opened the pull request:
## Involved packages

 - snpeff 

## Details
<!-- 
If you would like to continue to work with us, we will invite you as a member of this organization.
Fill the detials using x for what you've done. For example
- [x] Would like to continue to work with us
-->
- [x] Tested in the local machine (largest 16G RAM) and it is passed without any issue
- [x] Provide New Package
- [x] Fix the Packages
  - [x] PKGBUILD
  - [x] lilac.yaml
  - [ ] lilac.py
- [ ] Would like to continue to work with us

## Additional Note
the package provides both snpEff and SnpSift binaries.

starsareintherose commented:
I would suggest you complete the lilac.yaml just like what is written in tnt.

If this is package is maintained by yourself on AUR, add `update_aur_repo()` function.

For java package, better to compile from source unless it's difficult. I am still trying to fix current java applications. see my efforts on https://github.com/BioArchLinux/Packages/issues/85

starsareintherose commented:
try mvn package and see target or somewhere.

kbipinkumar commented:
Yes the package is maintained by me in AUR. I can make the modifications as you suggested. I will add bioarchlinux bot as co-maintainer as well. However, regarding compiling java packages from source is out of my skill level as of now. On 01-Dec-2022, at 3:42 PM, ***@***.*** wrote:﻿
I would suggest you complete the lilac.yaml just like what is written in tnt.
If this is package is maintained by yourself on AUR, add update_aur_repo() function.
For java package, better to compile from source unless it's difficult. I am still trying to fix current java applications. see my efforts on #85

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

starsareintherose commented:
just try `mvn package` command to check if it will fail or not. Usually command is `mvn package` (pom.xml) or `ant dist` (build.xml)

kbipinkumar commented:
> I would suggest you complete the lilac.yaml just like what is written in tnt.
> 
> If this is package is maintained by yourself on AUR, add `update_aur_repo()` function.
> 
> For java package, better to compile from source unless it's difficult. I am still trying to fix current java applications. see my efforts on #85



> just try `mvn package` command to check if it will fail or not. Usually command is `mvn package` (pom.xml) or `ant dist` (build.xml)

okay. i will give it a try

hubutui commented on BioArchLinux/snpeff/PKGBUILD:
```diff
@@ -0,0 +1,33 @@
+pkgdesc="Genetic variant annotation and effect prediction toolbox. https://doi.org/10.4161/fly.19695"
+arch=('any')
+url="https://pcingola.github.io/SnpEff/"
+provides=(""${pkgname%-pkgver}"")
```
A pkg named `snpeff` of course provides `snpeff`, aka itself. So this line is not needed.

hubutui commented on BioArchLinux/snpeff/PKGBUILD:
```diff
@@ -0,0 +1,33 @@
+source=(
+  snpEff.sh
+  SnpSift.sh
+  https://snpeff.blob.core.windows.net/versions/snpEff_latest_core.zip
```
It's not a good version, `vercmp 5.1d 5.1` returns `-1`, which means `5.1d < 5.1`. The upstream might need to use proper version number.

kbipinkumar replied:
> I recommend to use `https://github.com/pcingola/SnpEff/archive/refs/tags/v${pkgver}.tar.gz`, and compile from source.
> 
> BTW, the latest version is `5.1d`, see https://github.com/pcingola/SnpEff/releases/tag/v5.1d

That seems to be source code release. However the project main page as well as binaries lists version as 5.1

sukanka replied:
> That seems to be source code release. 

Whenever possible, compile from source.

> However the project main page as well as binaries lists version as 5.1

This means the project main page is out-of-date and should be updated.

BTW, as the version number is not properly set, you may file an issue in the upstream first.

hubutui replied:
why not confirm with upstream? It would be a problem if the source and the binary release did not match.

kbipinkumar commented:
> You might check also this [doc](https://pcingola.github.io/SnpEff/download/#building-from-the-source).

so according to the doc biojava-core and biojava-structure are also dependencies for building snpeff from source

kbipinkumar commented:
> just try `mvn package` command to check if it will fail or not. Usually command is `mvn package` (pom.xml) or `ant dist` (build.xml)

so according to the doc biojava-core and biojava-structure are also dependencies for building snpeff from source

starsareintherose commented:
> > just try `mvn package` command to check if it will fail or not. Usually command is `mvn package` (pom.xml) or `ant dist` (build.xml)
> 
> so according to the doc biojava-core and biojava-structure are also dependencies for building snpeff from source

already in pom.xml https://github.com/pcingola/SnpEff/blob/master/pom.xml#L102

hubutui commented:
you might also make sure the files' permission is set to `644`. I don't think they need `x` permission.

kbipinkumar commented:
> > > just try `mvn package` command to check if it will fail or not. Usually command is `mvn package` (pom.xml) or `ant dist` (build.xml)
> > 
> > 
> > so according to the doc biojava-core and biojava-structure are also dependencies for building snpeff from source
> 
> already in pom.xml https://github.com/pcingola/SnpEff/blob/master/pom.xml#L102

i have been trying to compile this package from sources. however, the sources for  two utilities of the package SnpEff and SnpSift are hosted in two different github repositories as per the [documentation](https://pcingola.github.io/SnpEff/download/#getting-the-source) from upstream as shown below.
```
# Get SnpEff
git clone https://github.com/pcingola/SnpEff.git

# Get SnpSift as well
git clone https://github.com/pcingola/SnpSift.git
```
i managed to write a [`PKGBUILD` ](https://gist.github.com/kbipinkumar/24a3b653e5a79aa41562902c221b74dc)that can download source tar files from both repositories. 
https://gist.github.com/kbipinkumar/24a3b653e5a79aa41562902c221b74dc
with this PKGBUILD i was able to successfully compile and package SnpEff. however while compiling SnpSift i am encountering following error.
```
[INFO] ------------------------< org.snpsift:SnpSift >-------------------------
[INFO] Building SnpSift 5.1
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.214 s
[INFO] Finished at: 2022-12-16T18:21:17+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project SnpSift: Could not resolve dependencies for project org.snpsift:SnpSift:jar:5.1: Failed to collect dependencies at org.snpeff:SnpEff:jar:5.1: Failed to read artifact descriptor for org.snpeff:SnpEff:jar:5.1: Could not transfer artifact org.snpeff:SnpEff:pom:5.1 from/to maven-default-http-blocker (http://0.0.0.0/): Blocked mirror for repositories: [typesafe (http://repo.typesafe.com/typesafe/releases/, default, releases+snapshots), hadoop-bam (http://hadoop-bam.sourceforge.net/maven/, default, releases+snapshots)] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```
i would be grateful if anyone could test it out in their machines to see if the error is coming up. also any further help in troubleshooting is extremely appreciated 🙏

sukanka commented:
> i managed to write a [`PKGBUILD` ](https://gist.github.com/kbipinkumar/24a3b653e5a79aa41562902c221b74dc)that can download source tar files from both repositories. https://gist.github.com/kbipinkumar/24a3b653e5a79aa41562902c221b74dc with this PKGBUILD i was able to successfully compile SnpEff and SnpSift. please take a look and let me know.

Looks fine to me, but there are minor issues you should pay attention to.


According to [Java_package_guidelines](https://wiki.archlinux.org/title/Java_package_guidelines#Dependencies), `java-environment` instead of `jdk-openjdk` should be used.

And since nothing is done in `prepare()`, it can be removed.

BTW, since the two components share the same version number, I now think `5.1` is better than `5.1d`. (`SnpSift` has no `5.1d` tag, so  `5.1` is fine.)

kbipinkumar commented:
> > i managed to write a [`PKGBUILD` ](https://gist.github.com/kbipinkumar/24a3b653e5a79aa41562902c221b74dc)that can download source tar files from both repositories. https://gist.github.com/kbipinkumar/24a3b653e5a79aa41562902c221b74dc with this PKGBUILD i was able to successfully compile SnpEff and SnpSift. please take a look and let me know.
> 
> Looks fine to me, but there are minor issues you should pay attention to.
> 
> According to [Java_package_guidelines](https://wiki.archlinux.org/title/Java_package_guidelines#Dependencies), `java-environment` instead of `jdk-openjdk` should be used.
> 
> And since nothing is done in `prepare()`, it can be removed.
> 
> BTW, since the two components share the same version number, I now think `5.1` is better than `5.1d`. (`SnpSift` has no `5.1d` tag, so `5.1` is fine.)

i have the modifications as suggested as well as adding shell wrapper for jar files. please check and let me know if anything else is to be done.

sukanka commented on BioArchLinux/snpeff/lilac.yaml:
```diff
@@ -5,7 +5,6 @@ build_prefix: extra-x86_64
 post_build_script: |
   git_add_files('PKGBUILD')
   git_commit()
-  update_aur_repo()
```
May I know why you removed this line, by mistake or intended?

kbipinkumar replied:
just as a caution. once the compiled package is tested and found fine we  can add update_aur_repo() back. i know for sure that AUR version is working fine so dont want to cause any issues with it for the time being. 🤞

sukanka replied:
> just as a caution. My intention was once the compiled package is tested and found fine we can add update_aur_repo() back. i know for sure that AUR version is working fine so don't want to cause any issues with it for the time being. crossed_fingers
> 
Good idea, you can test the PKGBUILD following the steps
1.  change to the directory that contains PKGBUILD of this pkg, 
2. then use `bioarchlinux-x86_64-build` to build it. 
3. If it is ok, you can then test whether it works fine. 
4. If it works, please add `update_aur_repo()` back.

BTW, if it failed to be built, it won't be pushed to aur even if `update_aur_repo()` is present.

kbipinkumar replied:
i have compiled and installed the package using updated `PKGBUILD`. So far in my usage, the newly built binary seems to be working fine. so, it should be okay to add `update_aur_repo()` back.

hubutui commented:
1. `MIT` license is not listed in `licenses` pkg, so we should install it to `/usr/share/licenses/snpeff`.
2. `PKGBUILD`, `lilac.py`, `lilac.yaml` should be change to `644` permission.
3. maybe remove the suffix of file `snpEff.sh` and `snpSift.sh`? I'm not sure if all lower-case is preferred.

kbipinkumar commented:
> 5. Could you git rebase and squash all these commits into one?

while attempting to squash commits i am getting following error. 
```
error: could not apply f075eca39... Rewrote PKGBUILD to compile from source
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
**Could not apply f075eca39... Rewrote PKGBUILD to compile from source**
```
commit f075eca39 is the current PKGBUILD that compiles from source. 

any ideas?


if nothing works, this pull request can be closed and new one be opened as snpeff has already been added ot Bioarchlinux:master branch. the rewritten PKGBUILD cn be submitted as fresh PR

kbipinkumar commented:
> 5. Could you git rebase and squash all these commits into one?

while attempting to squash commits i am getting following error. 
```
error: could not apply f075eca39... Rewrote PKGBUILD to compile from source
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
**Could not apply f075eca39... Rewrote PKGBUILD to compile from source**
```
commit f075eca39 is the current PKGBUILD that compiles

hubutui commented:
I think you could update the `PKGBUILD` to the file we wanted, and `git add PKGBUILD`, and `git rebase --continue`, then finally `git push -f` to your fork.

hubutui commented:
Oh, it seems you fork this repo, and work on a branch `snpeff` which also merge `https://github.com/kbipinkumar/Packages_sub/commit/fe5f5ff899acd9b1602adaa7bed5d0bc3779868d` to the branch, which make it complicated.

It would be much easier to:
1. rebase your forked repo https://github.com/kbipinkumar/Packages_sub to this repo, which could be done as [the doc](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork?platform=linux) describe.
2. create another branch from the updated forked repo, update the files we need, commit, and push it to the new branch in your forked repo.
3. create another pull request.

Maybe you could use the same branch name in the forked repo, and then you don't need to create another pull request. If you choose this method, remember to back up your files before deleting the old branch so that you could create a new branch with the same name.

hubutui commented:
another tip, `LF` line terminator is preferred. It seems `lilac.yaml` is with CRLF line terminators.
